text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,284 @@
+---
+layout: default
+title: Fabric Private Chaincode
+nav_order: 3
+---
+ 
+- Feature Name: Fabric Private Chaincode (FPC)
+- Start Date: 2020-01-28)
+- RFC PR: (leave this empty)
+- Fabric Component: Core (fill me in with underlined fabric component, core, orderer/consensus and etc.)
+- Fabric Issue: (leave this empty)
+ 
+# Summary
+[summary]: #summary
+
+This RFC aims to introduce a new security feature for Hyperledger Fabric called Fabric Private Chaincode (FPC) that enhances data confidentiality for chaincodes by executing them in a Trusted Execution Environment (TEE). The first supported TEE platform is Intel SGX, which protects the confidentiality and the integrity of the chaincode execution.
+ 
+FPC is a framework to develop and run chaincodes with strong privacy guarantees.
+With FPC, a chaincode can process transaction arguments and state without revealing/exposing the contents to anybody, including the endorsing peers. That is, clients can encrypt the content of a transaction in a way only the FPC chaincode can decrypt and then process it. The resulting state updates are encrypted by default by the FPC chaincode with keys only known to the chaincode. 
+
+As a result, other parties (both clients and peers) cannot examine the state of the transactions either in the chaincode or on the ledger. Hence, FPC helps to protects sensitive data even from compromised endorsing peers and other unauthorised parties.
+FPC provides the capabilities to verify that a chaincode and the data are protected, that is, Fabric peers and clients can receive cryptographic assurance that the correct chaincode is being run without tampering inside the TEE by means of a hardware-based remote attestation. 
+
+This new model of trust for Fabric smart contracts makes possible high-stakes markets such as private voting systems and sealed-bid auctions; which aren’t supported by the existing model of privacy in Fabric because of the requirement for endorsing peers to be fully trusted for confidentiality.
+ 
+A prototype implementation of FPC is available as Hyperledger Lab on github (https://github.com/hyperledger-labs/fabric-private-chaincode).
```

Comment:
If a prototype is available, Did we compare the performance of vanilla Fabric v2.0 and the v2.0 with the enclave with a different number of chaincodes, etc... ?

I am asking this because the performance could degrade significantly with the usage of enclaves -- _sgx-perf: A Performance Analysis Tool for Intel SGX Enclaves, Middleware 2019_ https://www.ibr.cs.tu-bs.de/users/weichbr/papers/middleware2018.pdf It might be better to quantify the drop in performance in terms of percentage.

Reply:
We have a performance evaluation using the prototype for Fabric v1.4, however, I don't think that this is representative anymore for the current prototype code.

Please see the paper: 
Marcus Brandenburger, Christian Cachin, Rüdiger Kapitza, Alessandro Sorniotti: Blockchain and Trusted Computing: Problems, Pitfalls, and a Solution for Hyperledger Fabric. https://arxiv.org/abs/1805.08541

Reply:
20% overhead seems like the best case. I was expecting at least 50% drop in the performance based on this paper  https://www.ibr.cs.tu-bs.de/users/weichbr/papers/middleware2018.pdf Is enclave execution efficient?

Reply:
Overhead is quite use-case dependent.  One of the overheads is related to the number of enclave <-> non-enclave transitions you do (which in turn is proportional to the shim calls you do.   This can be greatly optimized by using the switchless/exitless techniques of doing OCALLS. While we currently do not do that yet, it is certainly something which we will explore once more use-cases are clear and performance becomes an issue. (Note with FPC you will require less redundancy than in "vanilla" fabric, so there are compensating performance benefits).